### PR TITLE
JS: Update `DiagnosticLocation` call to gracefully handle invalid locations

### PR DESCRIPTION
**Pull request by henrymercer**
This PR updates the JS autobuilder's use of the `DiagnosticLocation` class to take advantage of an internal change that more gracefully handles invalid locations.

### CODE REVIEW

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java [hunk 1]
```diff
@@ -685,7 +687,8 @@ public int compare(Path f1, Path f2) {
    * Like {@link #PATH_ORDERING} but for {@link File} objects.
    */
   public static final Comparator<File> FILE_ORDERING = new Comparator<File>() {
-    public int compare(File f1, File f2) {
+    @Override
+	public int compare(File f1, File f2) {
```
Comment (erik-krogh) [inline javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:691]: This indentation looks weird?

**Review by erik-krogh — APPROVED** [review]
LGTM 👍 

But I think some of the indentation looks weird.

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java [hunk 2]
```diff
@@ -1238,18 +1241,24 @@ private void doExtract(FileExtractor extractor, Path file, ExtractorState state)
...
             .setEndColumn(err.getPosition().getColumn() + 1) // convert from 0-based to 1-based
-            .build();
-        writeDiagnostics(msg, JSDiagnosticKind.PARSE_ERROR, diagLoc);
+            .build()
+            .getOk();
+        }
+        if (diagLoc.isPresent()) {
+          writeDiagnostics(msg, JSDiagnosticKind.PARSE_ERROR, diagLoc.get());
+        } else {
+          msg += "\n\nRelated file: " + file;
```
Comment (aibaars) [inline javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1259]: I'm not sure we should "leak" the folder hierarchy of the machine on which the analysis was run into the diagnostic message. 

We should probably include a `DiagnosticLocation` with just the file path if the `build()` method fails due to invalid position information.

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java [hunk 3]
```diff
@@ -1238,18 +1241,24 @@ private void doExtract(FileExtractor extractor, Path file, ExtractorState state)
...
-        // file, relative to the source root
-        DiagnosticLocation.Builder builder = DiagnosticLocation.builder();
+
+        Optional<DiagnosticLocation> diagLoc = Optional.empty();
         if (file.startsWith(LGTM_SRC)) {
-          builder = builder.setFile(file.subpath(LGTM_SRC.getNameCount(), file.getNameCount()).toString());
-        }
-        DiagnosticLocation diagLoc = builder
+          diagLoc = DiagnosticLocation.builder()
+            .setFile(file.subpath(LGTM_SRC.getNameCount(), file.getNameCount()).toString()) // file, relative to the source root
```
Comment (aibaars) [inline javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1248]: Taking the relative path with `LGTM_SRC` is not always right. `LGTM_SRC` is simply the value of the current work directory and may not actually be the "repository root", for example when a custom `source-root` is used or the working directory is changed.

Comment (henrymercer) [reply javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1246]: Thanks — this seems like a fairly ingrained assumption in the autobuilder, so in the interests of breaking work down, I'll leave this for now as it is, and open an issue on the JS repo with your comments.

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java (hunk 2 above)
Comment (henrymercer) [reply javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java]: > I'm not sure we should "leak" the folder hierarchy of the machine on which the analysis was run into the diagnostic message.

That seems reasonable, particularly for the status page.  Done.

> We should probably include a `DiagnosticLocation` with just the file path if the `build()` method fails due to invalid position information.

We're just checking startLine <= endLine, so my understanding is we should never fail here due to invalid position information.

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java (hunk 1 above)
Comment (henrymercer) [reply javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java]: Thanks, fixed.  Out of interest, how do you tend to edit these files?  There doesn't seem to be a formatter configured in Eclipse.

**Comment by henrymercer** [comment]
@aibaars @erik-krogh Would you mind taking another look?  Thanks!

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java [hunk 4]
```diff
@@ -1237,19 +1238,29 @@ private void doExtract(FileExtractor extractor, Path file, ExtractorState state)
...
             .setStartLine(err.getPosition().getLine())
             .setStartColumn(err.getPosition().getColumn() + 1) // convert from 0-based to 1-based
             .setEndLine(err.getPosition().getLine())
             .setEndColumn(err.getPosition().getColumn() + 1) // convert from 0-based to 1-based
-            .build();
-        writeDiagnostics(msg, JSDiagnosticKind.PARSE_ERROR, diagLoc);
+            .build()
+            .getOk();
```
Comment (aibaars) [inline javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1252]:
```
ql/javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1252: error: cannot find symbol
            .getOk();
            ^
  symbol:   method getOk()
  location: class DiagnosticLocation
```

Comment (henrymercer) [reply javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1252]: This is due to us changing the `DiagnosticLocation` API in the corresponding internal PR (backlinked above).

**Review by aibaars — APPROVED** [review]

**Pull request merged**